modules/data.remote/R/Lantrendr_AGB_prep.R

```diff
@@ -0,0 +1,143 @@
+#' Prepare Landtrendr AGB data for the SDA workflow.
+#'
+#' @param site_info Bety list of site info including site_id, lon, and lat.
+#' @param start_date Start date of SDA workflow.
+#' @param end_date End date of SDA workflow.
+#' @param time_points A vector contains each time point within the start and end date.
+#' @param AGB_input_dir Where the Landtrendr AGB data can be accessed.
+#' @param outdir Where the final CSV file will be stored.
+#' @param export_csv Decide if we want to export the CSV file.
+#' @param allow_download If data is missing, should we download the missing data?
+#' @param buffer buffer area to calculate the min var of AGB data.
+#' @param skip_buffer flag to skip calculating min var based on buffer area.
+#'
+#' @return A data frame containing AGB median and sd for each site and each time step.
+#' @export
+#'
+#' @examples
+#' @author Dongchen Zhang
+#' @importFrom magrittr %>%
+Landtrendr_AGB_prep <- function(site_info, start_date, end_date, time_points, 
+                     AGB_input_dir = "/projectnb/dietzelab/dongchen/Multi-site/download_500_sites/AGB", 
+                     outdir = NULL, export_csv = TRUE, allow_download = FALSE, buffer = NULL, skip_buffer = TRUE){
+  #if we export CSV but didn't provide any path
+  if(as.logical(export_csv) && is.null(outdir)){
+    PEcAn.logger::logger.info("If you want to export CSV file, please ensure input the outdir!")
+    return(0)
+  }
+  #Landtrendr AGB doesn't provide data after 2017.
+  time_points <- time_points[which(lubridate::year(time_points) < 2018)]
+  
+  #check if we have all AGB data downloaded, if not, download them
+  if(as.logical(allow_download)){
+    AGB_median_years <- as.numeric(gsub(".*?([0-9]+).*", "\\1", list.files(AGB_input_dir, pattern = "*median.tif")))
+    missing_years_median <- lubridate::year((time_points[which(!lubridate::year(time_points)%in%AGB_median_years)])) #for landtrendr AGB data, we only have data before 2018.
+    
+    #starting downloading
+    if(length(missing_years_median)>0){
+      if(getOption('timeout') < 3600) options(timeout=3600)#enable 1h download time
+      PEcAn.data.remote::download.LandTrendr.AGB(outdir = AGB_input_dir, product_dates = missing_years_median, run_parallel = FALSE)
+    }
+  }
+  
+  #grab previous data to see which site has incomplete observations, if so, download the site for the whole time period.
+  #if we have previous downloaded CSV file
+  if(file.exists(file.path(outdir, "AGB.csv")) && length(buffer)==0 && as.logical(skip_buffer)){
+    Previous_CSV <- as.data.frame(utils::read.csv(file.path(outdir, "AGB.csv")))
+    AGB_Output <- matrix(NA, length(site_info$site_id), 2*length(time_points)+1) %>% 
+      `colnames<-`(c("site_id", paste0(time_points, "_AbvGrndWood"), paste0(time_points, "_SD"))) %>% as.data.frame()#we need: site_id, agb, sd, target time point.
```

Comment:
Organizing the data as time, site, var, SD is what I thought we all agreed on in the design doc as it should make it much easier to process

Reply:
I'm OK with leaving this as is for the current PR so folks can finish up the current round of papers, but design should be refactored before we move on after that.

Reply:
got it.